Spatial interpolation functionality

ahaberlie opened the pull request:
Adds interpolation functionality and various mapping utilities to metpy.

Specifically, this update provides wrappers for:

"linear", "nearest", "cubic", or "rbf" from Scipy.interpolate.

And implements the following interpolation approaches:

"natural_neighbor", "barnes", or "cressman"

Various mapping utilities are included as supporting functions:

area
barnes_weights
barnes_point
cressman_weights
cressman_point
calc_kappa
circumcenter
circumcircle_radius
circumcircle_radius_2
distance
dist_2
find_natural_neighbors
find_nn_triangles_point
find_local_boundary
generate_grid
get_boundary_coords
get_point_count_within_r
get_points_within_r
get_xy_range
get_xy_steps
nn_point
order_edges
remove_nan_observations
remove_observations_below_value
remove_repeat_coordinates

ahaberlie commented:
As you can see, the examples randomly fail based on travis ci build resources, which has nothing to do with the actual notebook content.  We could try doubling the timeout.

dopplershift commented:
Looks like there are no tests to execute *anything* in `gridding_functions.py`.

Feel free to add a commit that bumps the timeout for executing notebooks (`.travis.yml` and `appveyor.yml`).

ahaberlie commented:
Missed a filename change when I changed the directory names from mapping to gridding.

test_map_functions was in place to test the functions in test_gridding_functions.

Updated.

dopplershift commented on metpy/gridding/tests/test_gridding_functions.py:
```diff
@@ -0,0 +1,160 @@
+
+
+def test_interpolate(test_coords):
+    r"""Tests main interpolate function"""
```
I really don't like the one gigantic test for all methods:

- Once the first assert fails, nothing else in the function runs
- There's a lot of very similar function calls

You can use pytest's parameterized decorator to do something better:
```python
interp_methods = ['natural_neighbor', 'cressman', 'barnes', 'linear', 'nearest', 'cubic', 'rbf']

@pytest.mark.parametrize('method', interp_methods)
def test_interpolate(test_coords, method):
    xp, yp = test_coords
    xp *= 10
    yp *= 10
    z = ....
    extra_kw = {}
    if method == 'cressman':
        extra_kw['search_radius'] = 200
    elif method == 'barnes':
        extra_kw['search_radius'] = 400

    _, _, img = interpolate(xp, yp, z, hres=10, interp_type=method, **extra_kw)

    truth = np.load(get_test_data('{0}_test.npz'.format(method)))['img']

    assert_array_almost_equal(truth, img)
```

This makes it much easier to update as any number of things change. It also converts to multiple tests, which is what this really is.

dopplershift commented on metpy/gridding/tests/test_interpolation.py:
```diff
@@ -0,0 +1,186 @@
+    r"""Tests natural neighbor interpolation function"""
+
+    xp = np.array([8, 67, 79, 10, 52, 53, 98, 34, 15, 58])
+    yp = np.array([24, 87, 48, 94, 98, 66, 14, 24, 60, 16])
```
I think you should use another fixture here for the coords. You might consider using one for the grid (`xg` and `yg`) as well. The coords one (since there are like 5 places in here using them) would probably solve your codacy problem as well.

dopplershift commented on metpy/gridding/tests/test_interpolation.py:
```diff
@@ -0,0 +1,186 @@
+    truth = np.load(get_test_data("cressman_r20_mn1.npz"))['img']
+
+    assert_array_almost_equal(truth, img)
+
```
I'm thinking another parameterized test would be good here.

dopplershift commented on metpy/gridding/tests/test_points.py:
```diff
@@ -0,0 +1,160 @@
+    south_truth = 0
+    west_truth = 0
+
+    assert_array_almost_equal(north_truth, bbox['north'])
```
These aren't actually arrays, so why use almost equal? In fact, they're integers, so why bother with the amost part? You could just put the coords in a dictionary and let pytest handle it for you:
```python
truth = dict(east=9, north=9, south=0, west=0)
assert bbox == truth
```

dopplershift commented on metpy/gridding/__init__.py:
```diff
@@ -0,0 +1,16 @@
+__all__ = []
+__all__.extend(gridding_functions.__all__)  # pylint: disable=undefined-variable
+__all__.extend(points.__all__)  # pylint: disable=undefined-variable
+__all__.extend(triangles.__all__)  # pylint: disable=undefined-variable
```
Do you really think we should be including everything in `points`, `triangles`, and `polygons` as main API functions? That means eventually we intend to have everything that's been `exporter.export`-ed as API's that will not change and are supported to use. Certainly I see some functions in there (like `dist_2`) that definitely aren't, and I'm not sure any of these are things I consider beyond internal details from a "Meteorology for Python" perspective.

dopplershift commented on metpy/gridding/triangles.py:
```diff
@@ -0,0 +1,376 @@
+            pt2 = tri.simplices[triangle][(i + 1) % 3]
+
+            if (pt1, pt2) in edges:
+                edges.remove((pt1, pt2))
```
According to code coverage, this line is never executed. Is there some algorithmic reason that the points will always be `(pt2, pt1)` in `edges`? If so, we could simplify here.

ahaberlie replied:
I will have to try to figure out how to make a case where this can happen--IIRC it was added for cases encountered when interpolating for the U.S.

dopplershift commented:
I still need to go through some of the code in more detail.

I'm not sure what to do with the verification notebooks because I'm not sure they're useful examples--but I absolutely want to keep them around. Might just need to move them in the future (after we merge).

The wind/point notebooks have some commented out code, as well as some gnarly exception handling that should be pulled out (in `station_test_data`). Also, why use `StationPlot` in the wind notebook to plot wind barbs? Cartopy/Matplotlib have the stand alone `barbs` plot method, so there's no need to go through the `StationPlot` infrastructure--it just clouds the interpolation demonstration for no need.

ahaberlie commented:
Verification notebooks: Is there a way to only include these for developers (like testdata?)

dopplershift commented:
*sigh* Codacy annoys me sometimes...
